### ISSUE TITLE
Flannel + ContainerD doesn't work on bare metal k8s 1.22+ Windows Server 2022

### ISSUE THREAD

**Issue by vitaliy-leschenko** [kubernetes-sigs/sig-windows-tools#181]
**Describe the bug**
After merge (#161) into master I tried it. Quick description: it doesn't work.
Pods can't start with error:
```
  Warning  FailedCreatePodSandBox  4m21s (x131 over 32m)  kubelet            (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "ab19f6ca7afabcc1ecc54e934fdba133070f34dcace6edac7a275fb4829775a6": plugin type="nat" name="nat" failed (add): error creating endpoint hcnCreateEndpoint failed in Win32: IP address is either invalid or not part of any configured subnet(s). (0x803b001e) {"Success":false,"Error":"IP address is either invalid or not part of any configured subnet(s). ","ErrorCode":2151350302} : endpoint config &{ ab19f6ca7afabcc1ecc54e934fdba133070f34dcace6edac7a275fb4829775a6_nat 61f413f1-5bbd-4a23-a954-aef07d4fc361  [] [{ 0}] { [default.svc.cluster.local svc.cluster.local cluster.local] [10.96.0.10] [ndots:5]} [{192.168.101.6 0.0.0.0/0 0}]  0 {2 0}}
```

**To Reproduce**
The first problem that I found in `flannel\start.ps1` that powershell command `Copy-Item` doesn't have parameter `-DestinationPath`. I change it to `-Destination`.

The next problem it is we have no permission to move cni folder. I change `Move-Item` to `Copy-Item`.

And the last problem that I found it is value for `--iface` when we try to start `flanneld.exe`. It was hardcoded to 10.1.0.5. I change it to `$managementIP` - maybe it is wrong value. 

After all of this chanes flannel pod started:
```
Starting flannel
I1103 20:33:48.543742    4444 alivpc_windows.go:22] AliVpc is not supported on this platform
I1103 20:33:48.543742    4444 awsvpc_windows.go:22] AWS VPC is not supported on this platform
I1103 20:33:48.543742    4444 gce_windows.go:45] GCE is not supported on this platform
I1103 20:33:48.543742    4444 ipip_windows.go:23] ipip is not supported on this platform
I1103 20:33:48.543742    4444 ipsec_windows.go:20] ipsec is not supported on this platform
I1103 20:33:48.543742    4444 tencentvpc_windows.go:22] TencentVpc is not supported on this platform
I1103 20:33:48.543742    4444 udp_windows.go:22] udp is not supported on this platform
I1103 20:33:48.544678    4444 main.go:456] Searching for interface using 192.168.101.6
I1103 20:33:48.548549    4444 main.go:533] Using interface with name vEthernet (Ethernet) and address 192.168.101.6
I1103 20:33:48.548618    4444 main.go:550] Defaulting external address to interface address (192.168.101.6)
I1103 20:33:48.565755    4444 kube.go:116] Waiting 10m0s for node controller to sync
I1103 20:33:48.565755    4444 kube.go:299] Starting kube subnet manager
I1103 20:33:49.569534    4444 kube.go:123] Node controller sync successful
I1103 20:33:49.569534    4444 main.go:254] Created subnet manager: Kubernetes Subnet Manager - k8s-windows-1
I1103 20:33:49.569534    4444 main.go:257] Installing signal handlers
I1103 20:33:49.569657    4444 main.go:392] Found network config - Backend type: vxlan
I1103 20:33:49.569657    4444 vxlan_windows.go:126] VXLAN config: Name=flannel.4096 MacPrefix=0E-2A VNI=4096 Port=4789 GBP=false DirectRouting=false
I1103 20:33:49.588491    4444 device_windows.go:102] Found existing HostComputeNetwork flannel.4096
I1103 20:33:49.600258    4444 main.go:315] Changing default FORWARD chain policy to ACCEPT
I1103 20:33:49.600563    4444 main.go:323] Wrote subnet file to /run/flannel/subnet.env
I1103 20:33:49.600563    4444 main.go:327] Running backend.
I1103 20:33:49.600563    4444 main.go:345] Waiting for all goroutines to exit
I1103 20:33:49.600563    4444 vxlan_network_windows.go:63] Watching for new subnet leases
```

kube-proxy - can't find c:/opt/cni/bin/host-local.exe. I copied cni folder from flannel.

So now all pods started and looks like OK.
```
PS C:\Users\v.leschenko> kubectl --context containerd -n kube-system get pods
NAME                                               READY   STATUS    RESTARTS      AGE
coredns-78fcd69978-fjkj8                           1/1     Running   1 (10d ago)   10d
coredns-78fcd69978-swxcc                           1/1     Running   1 (10d ago)   10d
etcd-k8s-linux-1.digiteum.com                      1/1     Running   1 (10d ago)   10d
kube-apiserver-k8s-linux-1.digiteum.com            1/1     Running   1 (10d ago)   10d
kube-controller-manager-k8s-linux-1.digiteum.com   1/1     Running   1 (10d ago)   10d
kube-flannel-ds-sclds                              1/1     Running   0             10d
kube-flannel-ds-windows-amd64-ghcrw                1/1     Running   0             34m
kube-proxy-qpq7j                                   1/1     Running   1 (10d ago)   10d
kube-proxy-windows-txh4s                           1/1     Running   0             24h
kube-scheduler-k8s-linux-1.digiteum.com            1/1     Running   1 (10d ago)   10d
```

After this I tried to deploy pods (just nanoserver with `ping -t localhost`) and it can't starts.
```
PS C:\Users\v.leschenko> kubectl --context containerd get pods
NAME                    READY   STATUS              RESTARTS   AGE
demo-85d57c4449-78gz4   0/1     ContainerCreating   0          41m
```

```
PS C:\Users\v.leschenko> kubectl --context containerd describe pods/demo-85d57c4449-78gz4
Events:
  Type     Reason                  Age                    From               Message
  ----     ------                  ----                   ----               -------
  Normal   Scheduled               34m                    default-scheduler  Successfully assigned default/demo-85d57c4449-78gz4 to k8s-windows-1
  Warning  FailedCreatePodSandBox  34m                    kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "221fc61386087918562b72cb8b1875faca25333ae8d35aabd225439abcb97791": plugin type="nat" name="nat" failed (add): error creating endpoint hcnCreateEndpoint failed in Win32: IP address is either invalid or not part of any configured subnet(s). (0x803b001e) {"Success":false,"Error":"IP address is either invalid or not part of any configured subnet(s). ","ErrorCode":2151350302} : endpoint config &{ 221fc61386087918562b72cb8b1875faca25333ae8d35aabd225439abcb97791_nat 61f413f1-5bbd-4a23-a954-aef07d4fc361  [] [{ 0}] { [default.svc.cluster.local svc.cluster.local cluster.local] [10.96.0.10] [ndots:5]} [{192.168.101.6 0.0.0.0/0 0}]  0 {2 0}}
```

**Expected behavior**
Pods start and working well.

**Kubernetes (please complete the following information):**
 - Windows Server version: 2022
 - Kubernetes Version: 1.22.2
 - CNI: flannel 0.14

**Additional context**
kubelet and api-server have argument `--feature-gates=WindowsHostProcessContainers=true`

to prepare windows node I used https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/kubeadm scripts.

**Comment by lippertmarkus** [comment]
@vitaliy-leschenko you need to remove NAT CNI config from `C:\etc\cni\net.d` in order to force using Flannel. With that the hostprocess examples work fine for me.

**Comment by jsturtevant** [comment]
Thanks for trying it out and providing feedback 👍 

> to prepare windows node I used https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/kubeadm scripts.

 those scripts setup additional network files and also add a few additional binaries like wins which are not needed any more.  I've tested these against containerd nodes that were built with image builder: https://github.com/kubernetes-sigs/image-builder/blob/master/docs/book/src/capi/windows/windows.md  @lippertmarkus  Is this how you ended up with NAT CNI configuration too?

Maybe I can update the Readme to make note of that the install scripts won't work.  I assume you are using containerd 1.6+?

>The first problem that I found in flannel\start.ps1 that powershell command Copy-Item doesn't have parameter -DestinationPath. I change it to -Destination.

Which line is this? what error are you getting?

>The next problem it is we have no permission to move cni folder. I change Move-Item to Copy-Item.

This should work. Did the folder exist before you tried to copy the item there?  Might be due to the above issue?

>And the last problem that I found it is value for --iface when we try to start flanneld.exe. It was hardcoded to 10.1.0.5. I change it to $managementIP - maybe it is wrong value.

This was just a miss from when I testing locally.  This should be the interface your traffic goes through on the node. 

> kube-proxy - can't find c:/opt/cni/bin/host-local.exe. I copied cni folder from flannel.

This sounds like an error due to the https://github.com/kubernetes-sigs/sig-windows-tools/blob/bf507644f5e04cf1621a7a4fe01e70768c96ea91/hostprocess/flannel/flanneld/start.ps1#L5 not working for you.

**Comment by lippertmarkus** [comment]
I also used `https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/kubeadm` scripts which sets up additional network files leading to that error.

I just published a vagrant setup here: https://github.com/lippertmarkus/vagrant-k8s-win-hostprocess 
Probably most interesting for you are the setup scripts:
- controlplane: https://github.com/lippertmarkus/vagrant-k8s-win-hostprocess/blob/main/setup-scripts/controlplane.sh
- windows node: https://github.com/lippertmarkus/vagrant-k8s-win-hostprocess/blob/main/setup-scripts/winworker.ps1

[Here](https://github.com/lippertmarkus/vagrant-k8s-win-hostprocess/blob/main/setup-scripts/winworker.ps1#L18-L21) I'm deleting all networking config before joining the cluster (and Calico CNI configs are deployed)

**Comment by vitaliy-leschenko** [comment]
> I assume you are using containerd 1.6+?

Sure. 1.6.0 beta 1
> > The first problem that I found in flannel\start.ps1 that powershell command Copy-Item doesn't have parameter -DestinationPath. I change it to -Destination.
> 
> Which line is this? what error are you getting?

This line:
https://github.com/kubernetes-sigs/sig-windows-tools/blob/bf507644f5e04cf1621a7a4fe01e70768c96ea91/hostprocess/flannel/flanneld/start.ps1#L5

```
I have got error like this:
PS C:\Users\v.leschenko> Move-Item -Path a -DestinationPath b
Move-Item : A parameter cannot be found that matches parameter name 'DestinationPath'.
At line:1 char:19
+ Move-Item -Path a -DestinationPath b
+                   ~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Move-Item], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.Commands.MoveItemCommand
```

> 
> > The next problem it is we have no permission to move cni folder. I change Move-Item to Copy-Item.
> 
> This should work. Did the folder exist before you tried to copy the item there? Might be due to the above issue?
> 
> > And the last problem that I found it is value for --iface when we try to start flanneld.exe. It was hardcoded to 10.1.0.5. I change it to $managementIP - maybe it is wrong value.
> 
> This was just a miss from when I testing locally. This should be the interface your traffic goes through on the node.
> 
> > kube-proxy - can't find c:/opt/cni/bin/host-local.exe. I copied cni folder from flannel.
> 
> This sounds like an error due to the
> 
> https://github.com/kubernetes-sigs/sig-windows-tools/blob/bf507644f5e04cf1621a7a4fe01e70768c96ea91/hostprocess/flannel/flanneld/start.ps1#L5
> 
> not working for you.


As I understand the problem is how to prepare windows node to k8s join.

**Comment by vitaliy-leschenko** [comment]
> @vitaliy-leschenko I was running into the same issue with Calico. You need to remove NAT CNI config from `C:\etc\cni\net.d` in order to force using Flannel. With that the hostprocess examples work fine for me.

I tried to switch to Calico. And after remove `c:\etc\cni\net.d\0-containerd-nat.json` that was created in `Install-Containerd.ps1` script pods started.

**Comment by jsturtevant** [comment]
>     container_annotations = ["microsoft.com/hostprocess-container"]
>   pod_annotations = ["microsoft.com/hostprocess-container"]

This annotations should not be needed if you are on k8s 1.22+ (with the alpha feature enabled) and containerd 1.6+.  I would recommend not having them in the config as they would allow a user to bypass any policies you might put in place that check against a the pod specs `securityContext.windowsOptions.hostProcess`

**Comment by jsturtevant** [comment]
looks like I need to revisit the flannel implementation a bit.  Thanks for all the feedback and glad you got it working with calico!

/assign

**Comment by jsturtevant** [comment]
Initial issues should be fixed by #194.  lets start a new issue if not.
